ee/tabby-db/src/cache.rs

```diff
@@ -35,3 +36,52 @@ impl<T> Cache<T> {
         }
     }
 }
+
+pub struct TimedKeyedCache<K, V> {
+    timeout: Duration,
+    values: RwLock<HashMap<K, (DateTime<Utc>, V)>>,
+}
+
+impl<K, V> TimedKeyedCache<K, V> {
+    pub fn new(timeout: Duration) -> Arc<Self>
+    where
+        K: Send + Sync + 'static,
+        V: Send + Sync + 'static,
+    {
+        let cache = TimedKeyedCache {
+            timeout,
+            values: Default::default(),
+        };
+        let arc = Arc::new(cache);
+        let clone = arc.clone();
+        tokio::spawn(async move {
+            loop {
+                tokio::time::sleep(timeout.to_std().unwrap()).await;
+                let mut values = clone.values.write().await;
+                let now = Utc::now();
+                values.retain(|_k, (time, _v)| now - *time > clone.timeout);
+            }
+        });
+        arc
+    }
+
+    pub async fn get_or_refresh<F, E>(&self, key: K, refresh: impl Fn(K) -> F) -> Result<V, E>
```

Comment:
can interface being simplified into

```
get_or_refresh(&self, key: K, refresh: ..., timeout: Duration) {
   ...
}
```

In implementation, compare against `last_computed + timeout` with `now`. if expired, recompute and update last_computed

Reply:
Timeout is needed to auto-clear the cache periodically. Otherwise we have a potential memory leak